File: src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java
package frc.team2412.robot.subsystems;

import com.ctre.phoenix.motorcontrol.can.TalonFX;
import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX;
import edu.wpi.first.wpilibj.DoubleSolenoid;
import edu.wpi.first.wpilibj.DoubleSolenoid.Value;
import edu.wpi.first.wpilibj2.command.SubsystemBase;

import com.revrobotics.ColorMatch;
import com.revrobotics.ColorMatchResult;
import com.revrobotics.ColorSensorV3;

public class IntakeSubsystem extends SubsystemBase {

    // Constants

    public static class IntakeConstants {
        public static String teamColor = "placeholder";
        public static final double INTAKE_SPEED = 1;

    }

    // Enums

    public enum IntakeMotorState {
        IN, OUT, STOPPED;
    }

    public enum IntakeSolenoidState{
        EXTENDED, RETRACTED;
    }
    
    
    public enum IntakeState{
        EXTEND(DoubleSolenoid.Value.kForward),
        RETRACT(DoubleSolenoid.Value.kReverse);

        public DoubleSolenoid.Value value;

        private IntakeState(DoubleSolenoid.Value value) {
            this.value = value;
        }
    }
    


    // Define Hardware

    private final WPI_TalonFX motor1;

    private final WPI_TalonFX motor2;

    private final DoubleSolenoid solenoid1;

    //private final DoubleSolenoid solenoid2;

    //private final ColorSensorV3 colorSensor = new colorSensorV3(i2cPort);

    private IntakeMotorState intakeMotorState;

    private IntakeSolenoidState intakeSolenoidState;

    // CONSTRUCTOR!
    public IntakeSubsystem(WPI_TalonFX intakeMotor1, WPI_TalonFX intakeMotor2, DoubleSolenoid intakeSolenoid1) {
        this.motor1 = intakeMotor1;
        this.motor2 = intakeMotor2;

        this.motor2.setInverted(true);

        this.solenoid1 = intakeSolenoid1;
        //this.colorSensor = intakeColorSensor;
        intakeSolenoidState = IntakeSolenoidState.RETRACTED;
        intakeMotorState = IntakeMotorState.STOPPED;
        setName("IntakeSubsystem");
    }

    // Methods.

    public boolean checkMotorState() {
        return true;
    }

    public void intakeIn() {
        motor1.set(IntakeConstants.INTAKE_SPEED);
        motor2.set(IntakeConstants.INTAKE_SPEED);
        intakeMotorState = IntakeMotorState.IN;
    }

    public void intakeOut() {
        motor1.set(-IntakeConstants.INTAKE_SPEED);
        motor2.set(-IntakeConstants.INTAKE_SPEED);
        intakeMotorState = IntakeMotorState.OUT;
    }

    public void intakeStop() {
        motor1.set(0);
        motor2.set(0);
        intakeMotorState = IntakeMotorState.STOPPED;
    }

    public void intakeExtend() {
        intakeSolenoidState = IntakeSolenoidState.EXTENDED;

        solenoid1.set(IntakeState.EXTEND.value);
    }
    
    public void intakeRetract() {
        intakeSolenoidState = IntakeSolenoidState.RETRACTED; 

        solenoid1.set(IntakeState.RETRACT.value);
    }

    public boolean colorIdentify() {
        return true;
    }
    
}

File: src/main/java/frc/team2412/robot/subsystem/IntakeSubsystem.kt
package frc.team2412.robot.subsystem

import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX
import com.revrobotics.ColorSensorV3
import edu.wpi.first.wpilibj.Solenoid
import edu.wpi.first.wpilibj2.command.SubsystemBase

public class IntakeSubsystem(
        intakeMotor1: WPI_TalonFX,
        intakeMotor2: WPI_TalonFX,
        intakeSolenoid1: Solenoid,
        intakeSolenoid2: Solenoid,
        intakeColorSensor: ColorSensorV3
) : SubsystemBase() {
    public companion object IntakeConstants {
        public const val teamColor = "placeholder"
    }

    public enum class IntakeMotorState {
        IN,
        OUT,
        STOPPED
    }

    public enum class IntakeSolenoidState {
        EXTENDED,
        RETRACTED
    }

    public val motor1 = intakeMotor1
    public val motor2 = intakeMotor2
    public val solenoid1 = intakeSolenoid1
    public val solenoid2 = intakeSolenoid2
    public val colorSensor = intakeColorSensor

    private var intakeMotorState = IntakeMotorState.IN
    private var intakeSolenoidState = IntakeSolenoidState.RETRACTED

    public fun checkMotorState(): Boolean {
        return true
    }

    public fun intakeIn() {
        intakeMotorState = IntakeMotorState.IN
    }

    public fun intakeOut() {
        intakeMotorState = IntakeMotorState.OUT
    }

    public fun intakeStop() {
        intakeMotorState = IntakeMotorState.STOPPED
    }

    public fun intakeExtend() {
        intakeSolenoidState = IntakeSolenoidState.EXTENDED
    }

    public fun intakeRetract() {
        intakeSolenoidState = IntakeSolenoidState.RETRACTED
    }

    public fun colorIdentify(): Boolean {
        return true
    }
}
